File: cli/internal/connectapi/read_state.go
package connectapi

import (
	"context"
	"errors"

	"connectrpc.com/connect"
	"google.golang.org/protobuf/types/known/timestamppb"
	"hmans.de/chatto/internal/core"
	apiv1 "hmans.de/chatto/internal/pb/chatto/api/v1"
)

func (s *roomService) MarkRoomAsRead(ctx context.Context, req *connect.Request[apiv1.MarkRoomAsReadRequest]) (*connect.Response[apiv1.MarkRoomAsReadResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	result, err := s.api.core.ReadState().MarkRoomAsRead(ctx, caller.UserID, req.Msg.RoomId, req.Msg.UpToEventId)
	if err != nil {
		return nil, connectError(err)
	}

	resp := &apiv1.MarkRoomAsReadResponse{}
	if !result.LastReadAt.IsZero() {
		resp.LastReadAt = timestamppb.New(result.LastReadAt)
	}
	if !result.PreviousLastReadAt.IsZero() {
		resp.PreviousLastReadAt = timestamppb.New(result.PreviousLastReadAt)
	}
	return connect.NewResponse(resp), nil
}

func (s *threadService) MarkThreadAsRead(ctx context.Context, req *connect.Request[apiv1.MarkThreadAsReadRequest]) (*connect.Response[apiv1.MarkThreadAsReadResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}

	result, err := s.api.core.ReadState().MarkThreadAsRead(ctx, caller.UserID, req.Msg.RoomId, req.Msg.ThreadRootEventId, req.Msg.UpToEventId)
	if err != nil {
		return nil, connectError(err)
	}

	resp := &apiv1.MarkThreadAsReadResponse{}
	if !result.PreviousReadAt.IsZero() {
		resp.PreviousReadAt = timestamppb.New(result.PreviousReadAt)
	}
	return connect.NewResponse(resp), nil
}

func (s *roomService) AdvanceReadReceipt(ctx context.Context, req *connect.Request[apiv1.AdvanceReadReceiptRequest]) (*connect.Response[apiv1.AdvanceReadReceiptResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	result, err := s.api.core.ReadReceipts().Advance(ctx, caller.UserID, req.Msg.GetRoomId(), req.Msg.GetThreadRootEventId(), req.Msg.GetUpToEventId())
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.AdvanceReadReceiptResponse{Updated: result.Updated}), nil
}

func (s *roomService) GetReadReceiptSummaries(ctx context.Context, req *connect.Request[apiv1.GetReadReceiptSummariesRequest]) (*connect.Response[apiv1.GetReadReceiptSummariesResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	model := s.api.core.ReadReceipts()
	summaries, err := model.Summaries(ctx, caller.UserID, req.Msg.GetRoomId(), req.Msg.GetThreadRootEventId(), req.Msg.GetMessageEventIds())
	if errors.Is(err, core.ErrReadReceiptsDisabled) {
		return connect.NewResponse(&apiv1.GetReadReceiptSummariesResponse{Enabled: false}), nil
	}
	if err != nil {
		return nil, connectError(err)
	}
	response := &apiv1.GetReadReceiptSummariesResponse{Enabled: true, Summaries: make([]*apiv1.ReadReceiptSummary, 0, len(summaries))}
	for _, summary := range summaries {
		item := &apiv1.ReadReceiptSummary{
			MessageEventId: summary.MessageEventID,
			ReaderCount:    int32(summary.ReaderCount),
			PreviewUserIds: summary.PreviewUserIDs,
		}
		if !summary.LatestReadAt.IsZero() {
			item.LatestReadAt = timestamppb.New(summary.LatestReadAt)
		}
		response.Summaries = append(response.Summaries, item)
	}
	return connect.NewResponse(response), nil
}

func (s *roomService) ListReadReceiptReaders(ctx context.Context, req *connect.Request[apiv1.ListReadReceiptReadersRequest]) (*connect.Response[apiv1.ListReadReceiptReadersResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	model := s.api.core.ReadReceipts()
	readers, err := model.Readers(ctx, caller.UserID, req.Msg.GetRoomId(), req.Msg.GetThreadRootEventId(), req.Msg.GetMessageEventId())
	if errors.Is(err, core.ErrReadReceiptsDisabled) {
		return connect.NewResponse(&apiv1.ListReadReceiptReadersResponse{Enabled: false, Page: apiPageInfo(0, false)}), nil
	}
	if err != nil {
		return nil, connectError(err)
	}
	limit, offset := apiPagination(req.Msg.GetPage(), 50, 100)
	page, total, hasMore := apiSlicePage(readers, limit, offset)
	response := &apiv1.ListReadReceiptReadersResponse{
		Enabled: true,
		Readers: make([]*apiv1.ReadReceiptReader, 0, len(page)),
		Page:    apiPageInfo(total, hasMore),
	}
	users := &userService{api: s.api}
	for _, reader := range page {
		user, err := s.api.core.GetUser(ctx, reader.UserID)
		if err != nil {
			continue
		}
		apiUser, err := users.userSummary(ctx, user, nil)
		if err != nil {
			return nil, err
		}
		item := &apiv1.ReadReceiptReader{User: apiUser}
		if !reader.ReadAt.IsZero() {
			item.ReadAt = timestamppb.New(reader.ReadAt)
		}
		response.Readers = append(response.Readers, item)
	}
	return connect.NewResponse(response), nil
}

File: apps/frontend/src/lib/api-client/readState.ts
import { authHeaders, createTowkClient, handleAuthError } from './connect.js';
import { RoomService } from '@towk/api-types/api/v1/rooms_pb';
import type { User as APIUser } from '@towk/api-types/api/v1/users_pb';
import { ThreadService } from '@towk/api-types/api/v1/threads_pb';
import { protobufTimestampToISOString } from '$lib/protobufTimestamp';

export type ConnectAPIConfig = {
  serverId?: string;
  baseUrl: string;
  bearerToken: string | null;
  onAuthenticationRequired?: (serverId: string) => void;
};

export type MarkRoomAsReadResult = {
  lastReadAt: string | null;
  previousLastReadAt: string | null;
};

export type MarkThreadAsReadResult = {
  previousReadAt: string | null;
};

export type ReadReceiptSummary = {
  messageEventId: string;
  readerCount: number;
};

export type ReadReceiptReader = {
  id: string;
  login: string;
  displayName: string;
  avatarUrl: string | null;
  deleted: boolean;
  readAt: string | null;
};

export type ReadReceiptSummariesResult = {
  enabled: boolean;
  summaries: ReadReceiptSummary[];
};

export type ReadReceiptReadersResult = {
  enabled: boolean;
  readers: ReadReceiptReader[];
  totalCount: number;
  hasMore: boolean;
};

export function createReadStateAPI(config: ConnectAPIConfig) {
  const rooms = createTowkClient(RoomService, config);
  const threads = createTowkClient(ThreadService, config);
  const headers = () => authHeaders(config);
  return {
    async markRoomAsRead(input: {
      roomId: string;
      upToEventId?: string;
    }): Promise<MarkRoomAsReadResult> {
      try {
        const response = await rooms.markRoomAsRead(
          {
            roomId: input.roomId,
            upToEventId: input.upToEventId ?? ''
          },
          { headers: headers() }
        );
        return {
          lastReadAt: protobufTimestampToISOString(response.lastReadAt) ?? null,
          previousLastReadAt: protobufTimestampToISOString(response.previousLastReadAt) ?? null
        };
      } catch (err) {
        return handleAuthError(config, err);
      }
    },

    async markThreadAsRead(input: {
      roomId: string;
      threadRootEventId: string;
      upToEventId?: string;
    }): Promise<MarkThreadAsReadResult> {
      try {
        const response = await threads.markThreadAsRead(
          {
            roomId: input.roomId,
            threadRootEventId: input.threadRootEventId,
            upToEventId: input.upToEventId ?? ''
          },
          { headers: headers() }
        );
        return {
          previousReadAt: protobufTimestampToISOString(response.previousReadAt) ?? null
        };
      } catch (err) {
        return handleAuthError(config, err);
      }
    },

    async advanceReadReceipt(input: {
      roomId: string;
      threadRootEventId?: string | null;
      upToEventId: string;
    }): Promise<boolean> {
      try {
        const response = await rooms.advanceReadReceipt(
          {
            roomId: input.roomId,
            threadRootEventId: input.threadRootEventId ?? '',
            upToEventId: input.upToEventId
          },
          { headers: headers() }
        );
        return response.updated;
      } catch (err) {
        return handleAuthError(config, err);
      }
    },

    async getReadReceiptSummaries(input: {
      roomId: string;
      threadRootEventId?: string | null;
      messageEventIds: string[];
    }): Promise<ReadReceiptSummariesResult> {
      if (input.messageEventIds.length === 0) {
        return { enabled: true, summaries: [] };
      }
      try {
        const response = await rooms.getReadReceiptSummaries(
          {
            roomId: input.roomId,
            threadRootEventId: input.threadRootEventId ?? '',
            messageEventIds: input.messageEventIds
          },
          { headers: headers() }
        );
        return {
          enabled: response.enabled,
          summaries: response.summaries.map((summary) => ({
            messageEventId: summary.messageEventId,
            readerCount: summary.readerCount
          }))
        };
      } catch (err) {
        return handleAuthError(config, err);
      }
    },

    async listReadReceiptReaders(input: {
      roomId: string;
      threadRootEventId?: string | null;
      messageEventId: string;
      limit?: number;
      offset?: number;
    }): Promise<ReadReceiptReadersResult> {
      try {
        const response = await rooms.listReadReceiptReaders(
          {
            roomId: input.roomId,
            threadRootEventId: input.threadRootEventId ?? '',
            messageEventId: input.messageEventId,
            page: { limit: input.limit ?? 50, offset: input.offset ?? 0 }
          },
          { headers: headers() }
        );
        return {
          enabled: response.enabled,
          readers: response.readers.map((reader) => readReceiptReader(reader.user, reader.readAt)),
          totalCount: Number(response.page?.totalCount ?? response.readers.length),
          hasMore: response.page?.hasMore ?? false
        };
      } catch (err) {
        return handleAuthError(config, err);
      }
    }
  };
}

function readReceiptReader(
  user: APIUser | undefined,
  readAt: Parameters<typeof protobufTimestampToISOString>[0]
): ReadReceiptReader {
  return {
    id: user?.id ?? '',
    login: user?.login ?? '',
    displayName: user?.displayName || user?.login || '',
    avatarUrl: user?.avatarUrl ?? null,
    deleted: user?.deleted ?? false,
    readAt: protobufTimestampToISOString(readAt) ?? null
  };
}
